pkg/providers/instance/aksmachineinstance.go

```diff
@@ -0,0 +1,572 @@
+	}
+	instanceTypes = offerings.OrderInstanceTypesByPrice(instanceTypes, scheduling.NewNodeSelectorRequirementsWithMinValues(nodeClaim.Spec.Requirements...))
+
+	aksMachinePromise, err := p.beginCreateMachine(ctx, nodeClass, nodeClaim, instanceTypes, aksMachineName)
+	if err != nil {
+		// Clean up if creation fails.
+		if err := p.deleteMachine(ctx, aksMachineName); err != nil {
+			if !IsAKSMachineOrMachinesPoolNotFound(err) {
+				log.FromContext(ctx).Error(err, "failed to delete AKS machine after failed creation", "aksMachineName", aksMachineName)
+			}
+			// We don't return the cleanup error here, as we want to return the original error from beginCreateMachine
+		}
+		return nil, err
+	}
+
+	log.FromContext(ctx).Info("launched new AKS machine instance",
+		"aksMachineName", aksMachineName,
+		"instance-type", aksMachinePromise.InstanceType.Name,
+		"zone", aksMachinePromise.Zone,
+		"capacity-type", aksMachinePromise.CapacityType,
+	)
+
+	return aksMachinePromise, nil
+}
+
+func (p *DefaultAKSMachineProvider) Update(ctx context.Context, aksMachineName string, aksMachine armcontainerservice.Machine, etag *string) error {
+	var options *armcontainerservice.MachinesClientBeginCreateOrUpdateOptions
+	if etag != nil {
+		options = &armcontainerservice.MachinesClientBeginCreateOrUpdateOptions{
+			IfMatch: etag,
+		}
+	}
+
+	poller, err := p.azClient.aksMachinesClient.BeginCreateOrUpdate(ctx, p.clusterResourceGroup, p.clusterName, p.aksMachinesPoolName, aksMachineName, aksMachine, options)
+	if err != nil {
+		if IsAKSMachineOrMachinesPoolNotFound(err) {
+			// Can only be AKS machines pool not found.
+			// Suggestion: separate the util function to not cover more than needed?
+			return corecloudprovider.NewNodeClaimNotFoundError(fmt.Errorf("failed to begin update AKS machine %q: %w", aksMachineName, err))
+		}
+		return fmt.Errorf("failed to begin update AKS machine %q: %w", aksMachineName, err)
+	}
+	_, err = poller.PollUntilDone(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("failed to update AKS machine %q during LRO: %w", aksMachineName, err)
+	}
+	return nil
+}
+
+// ASSUMPTION: the AKS machine will be in the current p.aksMachinesPoolName. Otherwise need rework to pass the pool name in.
+func (p *DefaultAKSMachineProvider) Get(ctx context.Context, aksMachineName string) (*armcontainerservice.Machine, error) {
+	if p.aksMachinesPoolName == "" {
+		// Possible when this option field is not populated, which is not required when PROVISION_MODE is not aksmachineapi.
+		// But an AKS machine instance exists, whether added manually or from before switching PROVISION_MODE.
+		// So, we respond similarly to if AKS machines pool is not found.
+		return nil, corecloudprovider.NewNodeClaimNotFoundError(fmt.Errorf("failed to get AKS machine, AKS machines pool name is empty"))
+	}
+
+	// ASSUMPTION: AKS machines API accepts only AKS machine name.
```

Comment:
I've seen this assumption in a few places but not sure I follow what it means... it accepts the machine pool name and the machine, but is it really an assumption (does it deserve a comment?)

Feels like it's just a fact of the API?